elixir/apps/domain/lib/domain/auth/adapters/okta/api_client.ex

```diff
@@ -190,4 +202,52 @@ defmodule Domain.Auth.Adapters.Okta.APIClient do
   end
 
   defp parse_link_header(nil), do: nil
+
+  defp parse_headers_for_errors(headers) do
+    headers
+    |> Enum.find({}, fn {key, _val} -> key == "www-authenticate" end)
+    |> parse_error_header()
+  end
+
+  defp parse_error_header({"www-authenticate", errors}) do
+    String.split(errors, ",")
+    |> Enum.map(&String.trim/1)
+    |> Enum.filter(&String.starts_with?(&1, "error"))
+    |> Enum.map(&String.replace(&1, "\"", ""))
+    |> Enum.map(&String.split(&1, "="))
+    |> Enum.into(%{}, fn [k, v] -> {k, v} end)
+  end
+
+  defp parse_error_header(_) do
+    Logger.info("No www-authenticate header present")
+    %{"error" => "unknown", "error_message" => "no www-authenticate header present"}
+  end
+
+  defp fetch_latest_access_token(provider) do
+    access_token = provider.adapter_state["access_token"]
+
+    if access_token_active?(access_token) do
+      access_token
+    else
+      # Fetch provider from DB and return latest access token
+      {:ok, provider} = Domain.Auth.fetch_active_provider_by_id(provider.id)
+      provider.adapter_state["access_token"]
+    end
+  end
+
+  defp access_token_active?(nil), do: false
+  defp access_token_active?(""), do: false
+
+  defp access_token_active?(token) when is_binary(token) do
+    current_time = DateTime.utc_now()
+
+    with jwt <- JOSE.JWT.peek(token),
+         {:ok, timestamp_time} <- DateTime.from_unix(jwt.fields["exp"]),
+         :lt <- DateTime.compare(current_time, timestamp_time),
+         time_diff <- DateTime.diff(timestamp_time, current_time) do
+      time_diff >= 2 * 60
+    else
+      _ -> false
+    end
+  end
```

Comment:
Hm it may be a good idea to break this function into two that does the following:

1. Check that JWT is valid (peek) and `exp` is there
2. Check to ensure timestamp isn't expired